packages/discord.js/src/client/websocket/WebSocketShard.js

```diff
@@ -559,6 +583,48 @@ class WebSocketShard extends EventEmitter {
     }, 20_000).unref();
   }
 
+  /**
+   * Sets the RESUMED dispatch timeout.
+   * @param {number} [time] If set to -1, it will clear the resumed timeout
+   * @private
+   */
+  setResumedDispatchTimeout(time) {
+    if (time === -1) {
+      if (this.resumedDispatchTimeout) {
+        this.debug('Clearing the RESUMED dispatch timeout.');
+        clearTimeout(this.resumedDispatchTimeout);
+        this.resumedDispatchTimeout = null;
+      }
+      return;
+    }
+    this.debug('Setting a RESUMED dispatch timeout for 30s.');
+    this.resumedDispatchTimeout = setTimeout(() => {
+      this.debug('Did not receive RESUMED in time. Destroying and connecting again.');
+      this.destroy({ reset: false, closeCode: 4009 });
+    }, 30_000).unref();
+  }
+
+  /**
+   * Sets the READY dispatch timeout.
+   * @param {number} [time] If set to -1, it will clear the ready timeout
+   * @private
+   */
+  setReadyDispatchTimeout(time) {
+    if (time === -1) {
+      if (this.readyDispatchTimeout) {
+        this.debug('Clearing the READY dispatch timeout.');
+        clearTimeout(this.readyDispatchTimeout);
+        this.readyDispatchTimeout = null;
+      }
+      return;
+    }
+    this.debug('Setting a READY dispatch timeout for 20s.');
+    this.readyDispatchTimeout = setTimeout(() => {
+      this.debug('Did not receive READY in time. Destroying and connecting again.');
+      this.destroy({ reset: true, closeCode: 4009 });
+    }, 20_000).unref();
```

Comment:
I don't think we should call `.unref()` here, as this can potentially cause the process to exit early.

Also, why 20 seconds?

Reply:
> why 20 seconds?

As the other timeout was 20 seconds. READY should be faster than RESUMED as there are no packets sent in between